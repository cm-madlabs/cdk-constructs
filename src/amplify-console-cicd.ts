import * as amplify from '@aws-cdk/aws-amplify';
import { BuildSpec } from '@aws-cdk/aws-codebuild';
import * as cdk from '@aws-cdk/core';

export interface IAmplifyConsoleForSpaProps {
  readonly amplifyAppName: string;
  readonly isBasicAuth: boolean;
  readonly basicAuthSecretId: string;
  readonly basicAuthUserName: string;
  readonly githubTokenSecretId: string;
  readonly githubOwner: string;
  readonly githubRepo: string;
  readonly isAutoBuild: boolean;
  readonly preBuildCommands: string[];
  readonly buildCommands: string[];
  readonly baseDirectory: string;
}

export class AmplifyConsoleCiCd extends cdk.Construct {
  public readonly amplify: amplify.App;

  constructor(scope: cdk.Construct, id: string, props: IAmplifyConsoleForSpaProps) {
    super(scope, id);

    this.amplify = new amplify.App(this, id, {
      appName: props.amplifyAppName,
      buildSpec: BuildSpec.fromObject({
        version: 1,
        frontend: {
          phases: {
            preBuild: {
              commands: props.preBuildCommands,
            },
            build: {
              commands: props.buildCommands,
            },
          },
          artifacts: {
            baseDirectory: props.baseDirectory,
            files: ['**/*'],
          },
          cache: { paths: ['node_modules/**/*'] },
          customHeaders: [
            {
              pattern: '**/*',
              headers: [
                {
                  key: 'Strict-Transport-Security',
                  value: 'max-age=31536000; includeSubDomains',
                },
                { key: 'X-Frame-Options', value: 'DENY' },
                { key: 'X-XSS-Protection', value: '1; mode=block' },
                { key: 'X-Content-Type-Options', value: 'nosniff' },
              ],
            },
          ],
        },
      }),
      basicAuth: props.isBasicAuth ? amplify.BasicAuth.fromCredentials(
        props.basicAuthUserName,
        cdk.SecretValue.secretsManager(props.basicAuthSecretId),
      ): undefined,
      sourceCodeProvider: new amplify.GitHubSourceCodeProvider({
        oauthToken: cdk.SecretValue.secretsManager(props.githubTokenSecretId),
        owner: props.githubOwner,
        repository: props.githubRepo,
      }),
      autoBranchCreation: {
        autoBuild: props.isAutoBuild,
        patterns: ['master', 'main', 'feature/deploy*'],
        pullRequestPreview: false,
      },
      customRules: [
        {
          source: '</^[^.]+$|\\.(?!(css|gif|ico|jpg|js|png|txt|svg|woff|ttf|map|json)$)([^.]+$)/>',
          target: '/index.html',
          status: amplify.RedirectStatus.REWRITE,
        },
      ],
    });
  }
}