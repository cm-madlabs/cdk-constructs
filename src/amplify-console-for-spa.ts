import * as amplify from '@aws-cdk/aws-amplify';
import { BuildSpec } from '@aws-cdk/aws-codebuild';
import * as cdk from '@aws-cdk/core';

type IsBasicAuth = boolean;
type BasicAuthSecretId<T> = T extends true ? string : never;
type BasicAuthUserName<T> = T extends true ? string : never;

export interface AmplifyConsoleForSpaProps {
  amplifyAppName: string;
  isBasicAuth: IsBasicAuth;
  basicAuthSecretId: BasicAuthSecretId<IsBasicAuth>;
  basicAuthUserName: BasicAuthUserName<IsBasicAuth>;
  githubTokenSecretId: string;
  githubOwner: string;
  githubRepo: string;
  isAutoBuild: boolean;
  preBuildCommands: string[];
  buildCommands: string[];
  baseDirectory: string;
}

export class AmplifyConsoleForSpa extends cdk.Construct {
  public amplify: amplify.App;

  constructor(scope: cdk.Construct, id: string, props: AmplifyConsoleForSpaProps) {
    super(scope, id);

    this.amplify = new amplify.App(this, id, {
      appName: props.amplifyAppName,
      buildSpec: BuildSpec.fromObject({
        version: 1,
        frontend: {
          phases: {
            preBuild: {
              commands: props.preBuildCommands,
            },
            build: {
              commands: props.buildCommands,
            },
          },
          artifacts: {
            baseDirectory: props.baseDirectory,
            files: ['**/*'],
          },
          cache: { paths: ['node_modules/**/*'] },
          customHeaders: [
            {
              pattern: '**/*',
              headers: [
                {
                  key: 'Strict-Transport-Security',
                  value: 'max-age=31536000; includeSubDomains',
                },
                { key: 'X-Frame-Options', value: 'DENY' },
                { key: 'X-XSS-Protection', value: '1; mode=block' },
                { key: 'X-Content-Type-Options', value: 'nosniff' },
              ],
            },
          ],
        },
      }),
      basicAuth: props.isBasicAuth ? amplify.BasicAuth.fromCredentials(
        props.basicAuthUserName,
        cdk.SecretValue.secretsManager(props.basicAuthSecretId),
      ): undefined,
      sourceCodeProvider: new amplify.GitHubSourceCodeProvider({
        oauthToken: cdk.SecretValue.secretsManager(props.githubTokenSecretId),
        owner: props.githubOwner,
        repository: props.githubRepo,
      }),
      autoBranchCreation: {
        autoBuild: props.isAutoBuild,
        patterns: ['master', 'main', 'feature/deploy*'],
        pullRequestPreview: false,
      },
      customRules: [
        {
          source:
                        '</^[^.]+$|\\.(?!(css|gif|ico|jpg|js|png|txt|svg|woff|ttf|map|json)$)([^.]+$)/>',
          target: '/index.html',
          status: amplify.RedirectStatus.REWRITE,
        },
      ],
    });
  }
}