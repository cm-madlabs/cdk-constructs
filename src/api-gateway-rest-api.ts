import * as apiGateway from '@aws-cdk/aws-apigateway';
import * as logs from '@aws-cdk/aws-logs';
import * as cdk from '@aws-cdk/core';

export interface IApiGatewayRestApiProps {
  readonly apiName: string;
}

export class ApiGatewayRestApi extends cdk.Construct {
  public api: apiGateway.RestApi;

  constructor(scope: cdk.Construct, id: string, props: IApiGatewayRestApiProps) {
    super(scope, id);

    this.api = new apiGateway.RestApi(this, id, {
      restApiName: props.apiName,
      endpointConfiguration: {
        types: [apiGateway.EndpointType.REGIONAL],
      },
      deployOptions: {
        accessLogDestination: new apiGateway.LogGroupLogDestination(
          new logs.LogGroup(this, `${id}-ApiLogGroup`),
        ),
        accessLogFormat: apiGateway.AccessLogFormat.jsonWithStandardFields(),
        loggingLevel: apiGateway.MethodLoggingLevel.ERROR,
        metricsEnabled: true,
        stageName: 'v1',
        tracingEnabled: true,
      },
      defaultCorsPreflightOptions: {
        allowOrigins: apiGateway.Cors.ALL_ORIGINS,
        allowMethods: apiGateway.Cors.ALL_METHODS,
        allowCredentials: true,
      },
    });
  }
}