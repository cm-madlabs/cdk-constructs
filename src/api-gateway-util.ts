import {
  IntegrationResponse,
  MethodResponse,
  ConnectionType,
  LambdaIntegration,
  AuthorizationType,
  PassthroughBehavior,
  Resource, Method,
} from '@aws-cdk/aws-apigateway';
import { IFunction } from '@aws-cdk/aws-lambda';

/**
 * CDKでAPI Gatewayを定義するために、汎用的に使えるUtil関数
 */
export class ApiGatewayUtil {
  /**
   * REST 単一リソースを取得する GET メソッドを作成する
   * ex GET /users/{id}
   * @param resource
   * @param fn
   */
  public static createRestSingleGetMethod(resource: Resource, fn: IFunction): Method {
    const integration = new LambdaIntegration(
      fn,
      {
        proxy: false,
        connectionType: ConnectionType.INTERNET,
        passthroughBehavior: PassthroughBehavior.WHEN_NO_MATCH,
        requestTemplates: {
          'application/json': JSON.stringify({
            id: "$input.params('id')",
          }),
        },
        integrationResponses: ApiGatewayUtil.commonMethodIntegrationResponse({
          statusCode: '200',
          responseParameters: ApiGatewayUtil.corsHeaderIntegration(),
          responseTemplates: {
            'application/json': '$input.json(\'$\')',
          },
        }),
      },
    );
    return resource.addMethod('GET', integration, {
      authorizationType: AuthorizationType.CUSTOM,
      requestParameters: {
        'method.request.path.id': true,
      },
      methodResponses: ApiGatewayUtil.commonMethodOptionsResponse(),
    });
  }

  /**
   * Querystring で複数リソースを取得する GET メソッドを作成する
   * ex GET /users?name=xxx&limit=10&next_token=xxx
   * @param resource
   * @param fn
   */
  public static createQueryGetMethod(resource: Resource, fn: IFunction): Method {
    const integration = new LambdaIntegration(
      fn,
      {
        proxy: false,
        connectionType: ConnectionType.INTERNET,
        passthroughBehavior: PassthroughBehavior.WHEN_NO_MATCH,
        requestTemplates: {
          'application/json': `
                    #set($allParams = $input.params())
                    {
                        "body": {
                            #foreach($type in $allParams.keySet())
                                #if($type == "querystring")
                                    #set($params = $allParams.get($type))
                                    #foreach($paramName in $params.keySet())
                                        "$paramName" : "$util.escapeJavaScript($params.get($paramName))"
                                        #if($foreach.hasNext)
                                            ,
                                        #end
                                    #end
                                #end
                            #end
                        }
                    }
              `,
        },
        integrationResponses: ApiGatewayUtil.commonMethodIntegrationResponse({
          statusCode: '200',
          responseParameters: ApiGatewayUtil.corsHeaderIntegration(),
          responseTemplates: {
            'application/json': '$input.json(\'$\')',
          },
        }),
      },
    );
    return resource.addMethod('GET', integration, {
      methodResponses: ApiGatewayUtil.commonMethodOptionsResponse(),
    });
  }

  /**
   * API GatewayのCORSレスポンスヘッダーを返却する
   */
  public static corsResponseHeader(): { [destination: string]: boolean } {
    return {
      'method.response.header.Access-Control-Allow-Headers': true,
      'method.response.header.Access-Control-Allow-Methods': true,
      'method.response.header.Access-Control-Allow-Origin': true,
    };
  }

  /**
   * API Gatewayの統合のCORSレスポンスヘッダーを返却する
   */
  public static corsHeaderIntegration(): { [destination: string]: string } {
    return {
      'method.response.header.Access-Control-Allow-Headers':
          "'Origin,Content-Type,Authorization'",
      'method.response.header.Access-Control-Allow-Methods':
          "'GET,PUT,POST,PATCH,DELETE,OPTIONS'",
      'method.response.header.Access-Control-Allow-Origin': "'*'",
    };
  }

  /**
   * API Gatewayの統合レスポンス
   * @param defaultResponse
   */
  public static commonMethodIntegrationResponse(defaultResponse: IntegrationResponse): IntegrationResponse[] {
    return [
      defaultResponse,
      {
        statusCode: '400',
        responseParameters: this.corsHeaderIntegration(),

        responseTemplates: {
          'application/json': '$util.parseJson($input.json(\'$.errorMessage\'))',
        },
        selectionPattern: '^\\{\\"status\\":400.*',
      },
      {
        statusCode: '403',
        responseParameters: this.corsHeaderIntegration(),

        responseTemplates: {
          'application/json': '$util.parseJson($input.json(\'$.errorMessage\'))',
        },
        selectionPattern: '^\\{\\"status\\":403.*',
      },
      {
        statusCode: '404',
        responseParameters: this.corsHeaderIntegration(),

        responseTemplates: {
          'application/json': '$util.parseJson($input.json(\'$.errorMessage\'))',
        },
        selectionPattern: '^\\{\\"status\\":404.*',
      },
      {
        statusCode: '500',
        responseParameters: this.corsHeaderIntegration(),

        responseTemplates: {
          'application/json': '$util.parseJson($input.json(\'$.errorMessage\'))',
        },
        selectionPattern: '^\\{\\"status\\":500.*',
      },
    ];
  }

  /**
   * API Gatewayのメソッドレスポンスを定義
   */
  public static commonMethodOptionsResponse(): MethodResponse[] {
    return [
      {
        statusCode: '200',
        responseParameters: this.corsResponseHeader(),
      },
      {
        statusCode: '201',
        responseParameters: this.corsResponseHeader(),
      },
      {
        statusCode: '202',
        responseParameters: this.corsResponseHeader(),
      },
      {
        statusCode: '204',
        responseParameters: this.corsResponseHeader(),
      },
      {
        statusCode: '400',
        responseParameters: this.corsResponseHeader(),
      },
      {
        statusCode: '401',
        responseParameters: this.corsResponseHeader(),
      },
      {
        statusCode: '403',
        responseParameters: this.corsResponseHeader(),
      },
      {
        statusCode: '404',
        responseParameters: this.corsResponseHeader(),
      },
      {
        statusCode: '500',
        responseParameters: this.corsResponseHeader(),
      },
    ];
  }
}
