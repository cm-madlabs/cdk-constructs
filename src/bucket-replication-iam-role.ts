import {
  Effect,
  PolicyDocument,
  PolicyStatement,
  Role,
  ServicePrincipal,
} from '@aws-cdk/aws-iam';
import {
  Construct,
} from '@aws-cdk/core';

export interface IS3BucketReplicationIamRoleProps {
  /**
   * IAM Role名
   */
  readonly roleName?: string;

  /**
   * レプリケーション元のS3バケットArn
   */
  readonly baseBucketArn: string;

  /**
   * レプリケーション先のS3バケットArn
   */
  readonly sourceBucketArn: string;
}

/**
 * バケットレプリケーションを行うために、レプリケーション元に設定するIAMロールを作成するConstruct
 */
export class BucketReplicationIamRole extends Construct {
  public readonly role: Role;

  constructor(scope: Construct, id: string, props: IS3BucketReplicationIamRoleProps) {
    super(scope, id);

    this.role = new Role(this, 'Role', {
      roleName: props.roleName,
      assumedBy: new ServicePrincipal('s3.amazonaws.com'),
      inlinePolicies: {
        bucketReplicationPolicy: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                's3:ListBucket',
                's3:GetReplicationConfiguration',
                's3:GetObjectVersionForReplication',
                's3:GetObjectVersionAcl',
                's3:GetObjectVersionTagging',
                's3:GetObjectRetention',
                's3:GetObjectLegalHold',
              ],
              resources: [
                `${props.baseBucketArn}`,
                `${props.baseBucketArn}/*`,
                `${props.sourceBucketArn}`,
                `${props.sourceBucketArn}/*`,
              ],
            }),

            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                's3:ReplicateObject',
                's3:ReplicateDelete',
                's3:ReplicateTags',
                's3:ObjectOwnerOverrideToBucketOwner',
              ],
              resources: [
                `${props.baseBucketArn}/*`,
                `${props.sourceBucketArn}/*`,
              ],
            }),
          ],
        }),
      },
    });
  }
}
