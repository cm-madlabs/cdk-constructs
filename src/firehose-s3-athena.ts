import {
  Column,
  Database,
  DataFormat,
  Table,
} from '@aws-cdk/aws-glue';
import {
  ManagedPolicy,
  Role,
  ServicePrincipal,
} from '@aws-cdk/aws-iam';
import {
  CfnDeliveryStream,
} from '@aws-cdk/aws-kinesisfirehose';
import {
  Bucket,
} from '@aws-cdk/aws-s3';
import {
  Construct,
} from '@aws-cdk/core';

export interface IFirehoseS3Athena {
  /**
   * バケット名
   */
  readonly bucketName?: string;

  /**
   * IAMロール名
   */
  readonly roleName: string;

  /**
   * Firehose Delivery Stream名
   */
  readonly deliveryStreamName?: string;

  /**
   * Glueデータベース名
   */
  readonly databaseName: string;

  /**
   * Glueテーブル名
   */
  readonly tableName: string;

  /**
   * Athenaでクエリする対象となるS3バケットのパスprefix
   */
  readonly prefix: string;

  /**
   * Athenaでクエリするためのテーブルカラム
   */
  readonly columns: Column[];
}

/**
 * Kinesis Firehose経由でS3に保存してAthenaでクエリできるようにするConstruct
 */
export class FirehoseS3Athena extends Construct {
  public readonly bucket: Bucket;
  public readonly firehoseRole: Role;
  public readonly delivery: CfnDeliveryStream;
  public readonly database: Database;
  public readonly table: Table;

  constructor(scope: Construct, id: string, props: IFirehoseS3Athena) {
    super(scope, id);

    this.bucket = new Bucket(this, 'Bucket', {
      bucketName: props.bucketName,
    });

    this.firehoseRole = new Role(this, 'Role', {
      roleName: props.roleName,
      assumedBy: new ServicePrincipal('firehose.amazonaws.com'),
      managedPolicies: [
        ManagedPolicy.fromAwsManagedPolicyName('AmazonS3FullAccess'),
      ],
    });

    this.delivery = new CfnDeliveryStream(this, 'Delivery', {
      deliveryStreamName: props.deliveryStreamName,
      deliveryStreamType: 'DirectPut',
      s3DestinationConfiguration: {
        bucketArn: this.bucket.bucketArn,
        prefix: props.prefix,
        bufferingHints: {
          intervalInSeconds: 60,
          sizeInMBs: 5,
        },
        compressionFormat: 'GZIP',
        roleArn: this.firehoseRole.roleArn,
      },
    });

    this.database = new Database(this, 'Database', {
      databaseName: props.databaseName,
    });

    this.table = new Table(this, 'Table', {
      database: this.database,
      tableName: props.tableName,
      columns: props.columns,
      dataFormat: DataFormat.JSON,
      bucket: this.bucket,
      s3Prefix: props.prefix,
    });
  }
}
