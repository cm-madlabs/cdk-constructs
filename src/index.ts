export * from './amplify-console-cicd';
export * from './bucket-replication-iam-role';
export * from './firehose-s3-athena';
export * from './replication-destination-bucket';
