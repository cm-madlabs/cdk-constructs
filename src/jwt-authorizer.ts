import { Runtime, Function } from '@aws-cdk/aws-lambda';
import { NodejsFunction } from '@aws-cdk/aws-lambda-nodejs';
import * as cdk from '@aws-cdk/core';

export interface IJwtAuthorizer {
  functionName: string;
  jwksUrl: string;
  audience: string;
  issuer: string;
}

export class JwtAuthorizer extends cdk.Construct {
  public authorizer: Function;

  constructor(scope: cdk.Construct, id: string, props: IJwtAuthorizer) {
    super(scope, id);

    this.authorizer = new NodejsFunction(this, 'authorizer', {
      entry: './lib/handlers/jwt-authorizer.ts',
      handler: 'handler',
      runtime: Runtime.NODEJS_14_X,
      functionName: props.functionName,
      memorySize: 128,
      environment: {
        JWKS_URL: props.jwksUrl,
        AUDIENCE: props.audience,
        ISSUER: props.issuer,
      },
    });
  }
}