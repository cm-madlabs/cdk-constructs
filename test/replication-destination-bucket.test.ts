import { SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import { ReplicationDestinationBucket } from '../src';

describe('ReplicationDestinationBucket', () => {
  test('default', () => {
    const stack = new cdk.Stack();

    new ReplicationDestinationBucket(stack, 'test', {
      bucketName: 'test',
      baseIamRoleArn: 'test',
      destinationRoleArn: 'test',
    });

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });
});