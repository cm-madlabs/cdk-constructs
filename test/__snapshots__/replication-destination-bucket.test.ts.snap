// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReplicationDestinationBucket default 1`] = `
Object {
  "Resources": Object {
    "testD609C01D": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketName": "test",
        "VersioningConfiguration": Object {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "testPolicy42064509": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "testD609C01D",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:ReplicateObject",
                "s3:ReplicateDelete",
                "s3:ReplicateTags",
                "s3:ObjectOwnerOverrideToBucketOwner",
                "s3:GetObjectVersionTagging",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": "test",
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "testD609C01D",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
              "Sid": "1",
            },
            Object {
              "Action": Array [
                "s3:GetBucketVersioning",
                "s3:PutBucketVersioning",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": "test",
              },
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testD609C01D",
                  "Arn",
                ],
              },
              "Sid": "2",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
}
`;
